questdb-confstr/tests/tests.rs

```diff
@@ -0,0 +1,358 @@
+    assert_eq!(err.position(), 4);
+    assert_eq!(
+        err.to_string(),
+        "bad separator, expected ':' got '=' at position 4"
+    );
+}
+
+#[test]
+fn bad_key_value_separator_semicolon() {
+    let input = "http::host=localhost;port9000;";
+    let config = parse_conf_str(input);
+    assert!(config.is_err());
+    let err = config.unwrap_err();
+    assert_eq!(err.kind(), ErrorKind::BadSeparator(('=', ';')));
+    assert_eq!(err.position(), 29);
+    assert_eq!(
+        err.to_string(),
+        "bad separator, expected '=' got ';' at position 29"
+    );
+}
+
+#[test]
+fn bad_key_value_separator_colon() {
+    let input = "s3::host:localhost;port=9000;";
+    let config = parse_conf_str(input);
+    assert!(config.is_err());
+    let err = config.unwrap_err();
+    assert_eq!(err.kind(), ErrorKind::BadSeparator(('=', ':')));
+    assert_eq!(err.position(), 8);
+    assert_eq!(
+        err.to_string(),
+        "bad separator, expected '=' got ':' at position 8"
+    );
+}
+
+#[test]
+fn test_incomplete_key_no_value() {
+    let input = "http::host";
+    let config = parse_conf_str(input);
+    assert!(config.is_err());
+    let err = config.unwrap_err();
+    assert_eq!(err.kind(), ErrorKind::IncompleteKeyValue);
+    assert_eq!(err.position(), 10);
+    assert_eq!(
+        err.to_string(),
+        "incomplete key-value pair before end of input at position 10"
+    );
+}
+
+#[test]
+fn missing_trailing_semicolon() {
+    let input = "http::host=localhost;port=9000";
+    let config = parse_conf_str(input);
+    assert!(config.is_err());
+    let err = config.unwrap_err();
+    assert_eq!(ErrorKind::MissingTrailingSemicolon, err.kind());
+    assert_eq!(err.position(), 30);
+    assert_eq!(err.to_string(), "missing trailing semicolon at position 30");
+}
+
+#[test]
+fn escaped_semicolon_missing_trailing() {
+    let input = "http::host=localhost;port=9000;;";
+    let config = parse_conf_str(input);
+    assert!(config.is_err());
+    let err = config.unwrap_err();
+    assert_eq!(err.kind(), ErrorKind::MissingTrailingSemicolon);
+    assert_eq!(err.position(), 32);
+    assert_eq!(err.to_string(), "missing trailing semicolon at position 32");
+}
+
+#[test]
+fn escaped_semicolon() -> Result<(), ParsingError> {
+    let input = "FTP::HOSTS=abc.com;;def.com;;ghi.net;PORTS=9000;;8000;;7000;;;";
+    let config = parse_conf_str(input)?;
+    assert_eq!(config.service(), "ftp");
+    assert_eq!(config.get("hosts"), Some("abc.com;def.com;ghi.net"));
+    assert_eq!(config.get("ports"), Some("9000;8000;7000;"));
+    Ok(())
+}
+
+#[test]
+fn byte_not_char_position() {
+    let input = "p::n=静;:=42;";
+    let config = parse_conf_str(input);
+    assert!(config.is_err());
+    let err = config.unwrap_err();
+    assert_eq!(err.kind(), ErrorKind::ExpectedIdentifierNot(':'));
+    assert_eq!(err.position(), 9);
+    assert_eq!(
+        err.to_string(),
+        "expected identifier to start with ascii letter, not ':' at position 9"
```

Comment:
it looks like it's counting bytes and not chars. is that intentional? If I look at the input string then I don't see `:` at position 9.

Reply:
Yes, this _is_ intentional (and there's tests for it). It's because conting chars isn't really a thing, while you can slice the string at that point with the byte offset.

Reply:
I could track both, but for the few times anyone is ever going to use unicode chars and tell the difference it feels like an overkill.

Reply:
that's what I thought. but as a user - I would find it misleading. I don't run UTF-8 encoder in my head to tell where exactly is the position 9:)

I agree it's an edge-case and not worth spending much effort on it.

Reply:
Edge case, I think it's above good enough as-is.